### ISSUE TITLE
Proxy support

### ISSUE THREAD

**Issue by williamspindox** [spring-projects/spring-ai#251]
Hi, can you provide a proxy support during external spring-ai api calls? e.g. to openai or azure or huggingface..
Thanks, regards

**Comment by yaoyuandev** [comment]
Try command line option. I think it's more flexible.
[Java Networking and Proxies](https://docs.oracle.com/javase/8/docs/technotes/guides/net/proxies.html)

```bash
java -Dhttp.proxyHost=localhost -Dhttp.proxyPort=18080 -jar app.jar
```

**Comment by williamspindox** [comment]
Thanks, I put them in .mvn/jvm.config. To use a proxy even during debugging I solved it by using:

System.setProperty("https.proxyHost", "proxy.host");
System.setProperty("https.proxyPort", "proxy.port");
System.setProperty("https.proxyHost", "proxy.host");
System.setProperty("https.proxyPort", "proxy.port");

**Comment by williamspindox** [comment]
The workaround worked with "spring-ai-openai-spring-boot-starter", but not for "spring-ai-azure-openai-spring-boot-starter", I get:

java.net.UnknownHostException: xxx-openai-api.openai.azure.com
        at java.base/java.net.InetAddress$CachedAddresses.get(InetAddress.java:801) ~[na:na]
        at java.base/java.net.InetAddress.getAllByName0(InetAddress.java:1533) ~[na:na]
        at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1385) ~[na:na]
        at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1306) ~[na:na]
        at io.netty.util.internal.SocketUtils$9.run(SocketUtils.java:169) ~[netty-common-4.1.101.Final.jar:4.1.101.Final]
        at io.netty.util.internal.SocketUtils$9.run(SocketUtils.java:166) ~[netty-common-4.1.101.Final.jar:4.1.101.Final]
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:569) ~[na:na]
        at io.netty.util.internal.SocketUtils.allAddressesByName(SocketUtils.java:166) ~[netty-common-4.1.101.Final.jar:4.1.101.Final]        at io.netty.resolver.DefaultNameResolver.doResolveAll(DefaultNameResolver.java:50) ~[netty-resolver-4.1.101.Final.jar:4.1.101.Final]
        at io.netty.resolver.SimpleNameResolver.resolveAll(SimpleNameResolver.java:79) ~[netty-resolver-4.1.101.Final.jar:4.1.101.Final]
        at io.netty.resolver.SimpleNameResolver.resolveAll(SimpleNameResolver.java:71) ~[netty-resolver-4.1.101.Final.jar:4.1.101.Final]
        at io.netty.resolver.InetSocketAddressResolver.doResolveAll(InetSocketAddressResolver.java:73) ~[netty-resolver-4.1.101.Final.jar:4.1.101.Final]
        at io.netty.resolver.InetSocketAddressResolver.doResolveAll(InetSocketAddressResolver.java:31) ~[netty-resolver-4.1.101.Final.jar:4.1.101.Final]
        at io.netty.resolver.AbstractAddressResolver.resolveAll(AbstractAddressResolver.java:158) ~[netty-resolver-4.1.101.Final.jar:4.1.101.Final]
        at reactor.netty.transport.TransportConnector.doResolveAndConnect(TransportConnector.java:335) ~[reactor-netty-core-1.1.13.jar:1.1.13]
        at reactor.netty.transport.TransportConnector.lambda$connect$6(TransportConnector.java:165) ~[reactor-netty-core-1.1.13.jar:1.1.13]
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:132) ~[reactor-core-3.6.0.jar:3.6.0]
        at reactor.netty.transport.TransportConnector$MonoChannelPromise._subscribe(TransportConnector.java:638) ~[reactor-netty-core-1.1.13.jar:1.1.13]
        at reactor.netty.transport.TransportConnector$MonoChannelPromise.lambda$subscribe$0(TransportConnector.java:550) ~[reactor-netty-core-1.1.13.jar:1.1.13]
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173) ~[netty-common-4.1.101.Final.jar:4.1.101.Final]
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166) ~[netty-common-4.1.101.Final.jar:4.1.101.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470) ~[netty-common-4.1.101.Final.jar:4.1.101.Final]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569) ~[netty-transport-4.1.101.Final.jar:4.1.101.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[netty-common-4.1.101.Final.jar:4.1.101.Final]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.101.Final.jar:4.1.101.Final]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.101.Final.jar:4.1.101.Final]
        at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]

**Comment by gterminator** [comment]
I have the same problem. Setting environment variables, setting up Webclient Bean, setting up ClientHttpRequestFactory didn't work too.

**Comment by Darlingxxx-Paranoid** [comment]
When using Spring AI in a restricted enterprise network, I have the same problem:
1. UnknownHostException: Even with http.proxyHost set, the underlying Netty/OkHttp client tries to resolve DNS locally before connecting to the proxy.
2. Proxy Ignored: In dynamic model creation (using OpenAiApi.builder()), the global proxy settings or custom WebClient beans are often ignored.

**The Root Cause**
1. DNS Resolution: Reactive clients (Netty) often use an internal DNS resolver. In internal networks, local DNS can't resolve external domains like openai.com or dashscope.aliyuncs.com.
2. Decoupled Components: Spring AI's OpenAiApi can use either RestClient or WebClient. If you only configure one, the other will fail during specific operations (like streaming or connection testing).

**The Solution**
To solve this, you need to provide a Dual-Client Proxy Configuration to the OpenAiApi.builder(). We recommend using Apache HttpClient 5 for the synchronous RestClient to avoid Spring 6.2 deprecation issues.

Step 1: Add Dependency
Ensure you have the Apache HttpClient 5 in your pom.xml:
```
<dependency>
    <groupId>org.apache.httpcomponents.client5</groupId>
    <artifactId>httpclient5</artifactId>
</dependency>
```

Step 2: Implementation
Use the following pattern in your DynamicModelFactory or Configuration class:

```Java
@Service
public class AiModelFactory {

    @Value("${my.proxy.host}") String proxyHost;
    @Value("${my.proxy.port}") int proxyPort;
    @Value("${my.proxy.username}") String proxyUser;
    @Value("${my.proxy.password}") String proxyPass;

    /**
     * Solution for RestClient (Synchronous) - Fixes Spring 6.2 Deprecation
     */
    private RestClient.Builder getProxiedRestClientBuilder() {
        BasicCredentialsProvider credsProvider = new BasicCredentialsProvider();
        credsProvider.setCredentials(new AuthScope(proxyHost, proxyPort),
                new UsernamePasswordCredentials(proxyUser, proxyPass.toCharArray()));

        CloseableHttpClient httpClient = HttpClients.custom()
                .setProxy(new HttpHost(proxyHost, proxyPort))
                .setDefaultCredentialsProvider(credsProvider)
                .build();

        return RestClient.builder()
                .requestFactory(new HttpComponentsClientHttpRequestFactory(httpClient));
    }

    /**
     * Solution for WebClient (Asynchronous/Streaming) - Fixes UnknownHostException
     */
    private WebClient.Builder getProxiedWebClientBuilder() {
        HttpClient httpClient = HttpClient.create()
                .proxy(p -> p.type(ProxyProvider.Proxy.HTTP)
                        .host(proxyHost)
                        .port(proxyPort)
                        .username(proxyUser)
                        .password(s -> proxyPass));

        return WebClient.builder()
                .clientConnector(new ReactorClientHttpConnector(httpClient));
    }

    public ChatModel createChatModel(ModelConfigDTO config) {
        // Core: Inject BOTH builders into OpenAiApi
        OpenAiApi openAiApi = OpenAiApi.builder()
                .apiKey(config.getApiKey())
                .baseUrl(config.getBaseUrl())
                .restClientBuilder(getProxiedRestClientBuilder())
                .webClientBuilder(getProxiedWebClientBuilder()) 
                .build();

        return new OpenAiChatModel(openAiApi);
    }
}
```
